File: frontend/src/store/endpoints/progress.ts
import { api } from "@/store/api";
import type { DeckDto, ProgressDto, TutorBriefDto } from "@/types/api";

export const progressApi = api.injectEndpoints({
  endpoints: (build) => ({
    getProgress: build.query<ProgressDto, void>({
      query: () => "/progress",
      providesTags: ["Progress"],
    }),
    getDeck: build.query<DeckDto, void>({
      query: () => "/deck",
      providesTags: [{ type: "Card", id: "LIST" }, "Progress"],
    }),
    getTutorBrief: build.query<TutorBriefDto, void>({
      query: () => "/tutor-brief",
      providesTags: ["Progress"],
    }),
  }),
});

export const { useGetProgressQuery, useGetDeckQuery, useGetTutorBriefQuery } =
  progressApi;

File: backend/app/routers/progress.py
from datetime import timedelta

from fastapi import APIRouter, Request

from app.deps import DbDep, ProfileDep, SettingsDep
from app.models.progress import DeckDto, ProgressDto, TutorBriefDto
from app.services import progress, tutor
from app.services.registry import Providers

router = APIRouter()


def _providers(request: Request) -> Providers:
    providers: Providers = request.app.state.providers
    return providers


@router.get("/progress", response_model=ProgressDto)
async def get_progress(db: DbDep, profile: ProfileDep, settings: SettingsDep) -> ProgressDto:
    """Totals and the ten-day series. The struggling line staying flat is the whole story."""
    return await progress.progress(db, profile, timedelta(hours=settings.due_window_hours))


@router.get("/deck", response_model=DeckDto)
async def get_deck(db: DbDep, profile: ProfileDep, settings: SettingsDep) -> DeckDto:
    """Every card, newest first, with its lyric state: on (mastered), off (learning), miss (due)."""
    return await progress.deck(db, profile, timedelta(hours=settings.due_window_hours))


@router.get("/tutor-brief", response_model=TutorBriefDto)
async def get_tutor_brief(db: DbDep, profile: ProfileDep, request: Request) -> TutorBriefDto:
    """The weekly one-pager for a human tutor. Cached per week until new cards appear."""
    return await tutor.brief(db, profile, _providers(request))
